[CodeCompletion] Complete Swift only module name after 'import'

rintaro opened the pull request:
For each `ModuleLoader`, newly implement `collectVisibleTopLevelModuleNames()` which collects *visible* module names. It doesn't care about the actual content of the file though. So it doesn't guarantee the module is *loadable*.

rdar://problem/39392446

rintaro committed:
[CodeCompletion] Complete Swift only module name after 'import'

rdar://problem/39392446

rintaro committed:
[Serialize] Consolidate duplicated logics

- forEachTargetModuleBasename(): Iterate through target names to search
- forEachModuleSearchPath(): Iterate through search paths

rintaro committed:
[test] Add test cases for Swift module completion after 'import'

rintaro commented:
@swift-ci Please smoke test

rintaro commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -1753,35 +1753,29 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
+      if (ModuleName.str().startswith("_"))
+        continue;
+      if (ModuleName == Ctx.SwiftShimsModuleName)
+        continue;
```
Are there other module names we should hide? e.g. `SwiftOnoneSupport`

benlangmuir replied:
+1 to skipping OnoneSupport (looks like the name is in Strings.h).  Other modules we should skip: `__C`, `__C_Synthesized`, `Builtin`.  I'm guessing they are already hidden because they're not on the file system, but please add `CHECK-NOT`s for them.

brentdax replied:
`_InternalSwiftSyntaxParser` is another one we don't want to show. Should already be covered, but again, worth a test to make sure.

rintaro commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -44,6 +106,132 @@ SerializedModuleLoader::~SerializedModuleLoader() = default;
+    }
+    case SearchPathKind::Framework: {
+      // Look for:
+      // $PATH/{name}.framework/Modules/{name}.swiftmodule/{arch}.{extension}
```
Not sure we need to handle frameworks practically. If framework always have Clang module (`module.modulemap`) with corresponding module name, `ClangImporter` can find them.
@jrose-apple Is pure Swift only *framework* possible?

rintaro commented:
@swift-ci Please test

rintaro committed:
[ASTContext] Remove unused function and includes

rintaro commented:
@swift-ci Please test

benlangmuir commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -1753,35 +1753,22 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
-      auto MD = ModuleDecl::create(Ctx.getIdentifier(ModuleName), Ctx);
+      if (ModuleName == mainModuleName)
+        continue;
+      if (ModuleName.str().startswith("_"))
```
I'm not happy about this case, but since it's just been copied okay...

benlangmuir commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -44,6 +106,132 @@ SerializedModuleLoader::~SerializedModuleLoader() = default;
+
+  // Check whether target specific module file exists or not in given directory.
+  // $PATH/{arch}.{extension}
+  auto checkTargetFiles = [&](StringRef path) -> bool {
```
Is there any point in checking these instead of just detecting the presence of the .swiftmodule directory/file?  We aren't checking if the module is importable.

jrose-apple replied:
That'd be faster but might mean less common code between both code paths.

benlangmuir replied:
Good point. I'm okay with this if it's not a big perf issue and keeps the code simpler.

jrose-apple reviewed (commented):
I haven't taken a close look at the refactoring for SerializedModuleLoaderBase yet, but here's some initial comments. (Seems intricate…)

brentdax commented:
Initial reaction is that I like the module loading refactoring quite a bit, but I'll need to study it more to see if it preserves behavior.

rintaro commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -32,6 +32,68 @@ using namespace swift;
+
+/// Apply \c body for each target-specific module file base name to search.
+void forEachTargetModuleBasename(const ASTContext &Ctx,
+  llvm::function_ref<void(StringRef)> body) {
```
Ah, I forgot to apply `clang-format` to this block. Thanks!

rintaro committed:
[CodeCompletion] Hide 'SwiftOnoneSupport' from module name completion

beccadax reviewed (approved):
The `SerializedModuleLoaderBase` refactoring is *so* nice. I knew there was a small, sensible function struggling to escape from that mess.

rintaro commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -44,6 +106,132 @@ SerializedModuleLoader::~SerializedModuleLoader() = default;
+  // Apply \p body for each directory entry in \p dirPath.
+  auto forEachDirectoryEntryPath =
+      [&](StringRef dirPath, llvm::function_ref<void(StringRef)> body) {
+        std::error_code errorCode;
```
`errorCode` is handled in `for` condition. `for (; !errorCode && DI != End;...`

rintaro commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -44,6 +106,132 @@ SerializedModuleLoader::~SerializedModuleLoader() = default;
+      // Look for:
+      // (Darwin OS) $PATH/{name}.swiftmodule/{arch}.{extension}
+      // (Other OS)  $PATH/{name}.{extension}
+      bool requireTargetSpecificModule = Ctx.LangOpts.Target.isOSDarwin();
```
That seems to be a good idea. Let me try it in follow-ups.

rintaro commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -228,6 +401,7 @@ SerializedModuleLoaderBase::findModule(AccessPathElem moduleID,
 
     // We can only get here if all targetFileNamePairs failed with
     // 'std::errc::no_such_file_or_directory'.
+    auto normalizedTarget = getTargetSpecificModuleTriple(Ctx.LangOpts.Target);
```
This is diagnostics code path. I don't expect this code is going to be executed so often. Am I missing something?

brentdax replied:
If we end up at the `return None;` after this, we may still successfully load a module from a later search path. But I guess that only happens if there's a .swiftmodule directory *and* the directory doesn't contain any accessible .swiftmodule files, so it's not likely to happen repeatedly.

(Hmm, I don't think failed `#if canImport(...)` results are cached anywhere...unrelated, just something I noticed.)

rintaro replied:
Ah, that's true. Thank you for explanation!

rintaro replied:
Decided to use the first name from `forEachTargetModuleBasename()`

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[Serialization] Avoid to call getTargetSpecificModuleTriple() twice

Maintain an invariant that forEachTargetModuleBasename() iterates names
from most to least desiable.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

jrose-apple commented:
\*sigh\* I think the tri-state `Optional<bool>` ended up causing rdar://problem/50789839. I'll probably go through and replace it all with an enum.

rintaro commented:
@jrose-apple Could you elaborate? Did I make a mistake?

jrose-apple commented:
I'm not sure. This morning I'm thinking about it more and I think you *didn't* make a mistake, but I'm still taking this as evidence that the code isn't as easy-to-read as it could be.

jrose-apple commented:
(The issue is that ParseableInterfaceModuleLoader is failing but the compilation process overall is "succeeding", i.e. exiting with an exit code of 0.)
